File: Python/letCode/group_anagrams.py
"""Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once."""
import collections

def groupAnagrams( strs: list[str]) -> list[list[str]]:
        ans = collections.defaultdict(list)

        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord("a")] += 1
                
            ans[tuple(count)].append(s)
      
        return ans.values()
        
         
        
        
           


strs = ["eat","tea","tan","ate","nat","bat"]
print(groupAnagrams(strs))

File: Typescript/letCode/groupAnagrams.ts
import { isValidAnagram } from "./valid_anagram";

const groupAnagrams=(strs:string[])=>{
    const hash:{[key:string]:string[]}= {}
    strs.forEach((item)=>{
        let doesExist = false;
        Object.keys(hash).forEach((key)=>{
            if(isValidAnagram(item,key)){
                hash[key].push(item);
                doesExist = true;
            }
        });
        if(!doesExist){
            hash[item] = [item];
        }
    });
    console.log(hash)
    return [...Object.keys(hash).map((k)=>hash[k])]
}


console.log(groupAnagrams(['eat', 'ate', 'dog', 'pog']))
